runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateCache.java

```diff
@@ -116,222 +129,198 @@ private ForComputation(String computation) {
 
     /** Invalidate all cache entries for this computation and {@code processingKey}. */
     public void invalidate(ByteString processingKey, long shardingKey) {
-      synchronized (this) {
-        WindmillComputationKey key =
-            WindmillComputationKey.create(computation, processingKey, shardingKey);
-        for (StateId id : keyIndex.removeAll(key)) {
-          stateCache.invalidate(id);
-        }
-      }
+      WindmillComputationKey key =
+          WindmillComputationKey.create(computation, processingKey, shardingKey);
+      // By removing the ForKey object, all state for the key is orphaned in the cache and will
+      // be removed by normal cache cleanup.
+      keyIndex.remove(key);
     }
 
-    /** Returns a per-computation, per-key view of the state cache. */
-    public ForKey forKey(
+    /** Returns a per-computation, per-key, per-state-family view of the state cache. */
+    public ForKeyAndFamily forKey(
         WindmillComputationKey computationKey,
         String stateFamily,
         long cacheToken,
         long workToken) {
-      return new ForKey(computationKey, stateFamily, cacheToken, workToken);
+      ForKey forKey = keyIndex.get(computationKey);
```

Comment:
This could be racy if there are multiple accesses to the cache at the same time. I suspect it's ok because access is already synchronized by key (one keyed worked item active). However. please add a comment explaining the thread safety.

Reply:
done (there was one regarding this up by WindmillStateCache but I agree calling out here  makes sense too)